server/lib/notifications/email.ts

```diff
@@ -89,21 +112,24 @@ export const notify = {
     if (process.env.ONLY) {
       debug('ONLY set to ', process.env.ONLY, ' => skipping subscribers');
       const isTransactional = TransactionalActivities.includes(activity.type);
-      return emailLib.send(options?.template || activity.type, process.env.ONLY, activity.data, {
+      await emailLib.send(options?.template || activity.type, process.env.ONLY, activity.data, {
         ...options,
         isTransactional,
       });
     } else if (cleanUsersArray.length > 0) {
-      return Promise.all(
-        cleanUsersArray
-          // Filter out unsubscribed users
-          .filter(user => {
-            const isUnsubscribed = unsubscribed.some(unsubscribedUser => unsubscribedUser.id === user.id);
-            const isExcluded = options?.exclude?.includes(user.id) || false;
-            return !isUnsubscribed && !isExcluded;
-          })
-          .map(user => notify.user(activity, { ...options, unsubscribed, user })),
-      );
+      const queue = new PQueue({ concurrency: 50 });
+      for (const userOrUserId of cleanUsersArray) {
+        const isUserId = typeof userOrUserId === 'number';
+        const [userId, user] = isUserId ? [userOrUserId, null] : [userOrUserId.id, userOrUserId];
+        if (
+          !unsubscribed.some(unsubscribedUser => unsubscribedUser.id === userId) && // Unsubscribed
+          !options?.exclude?.includes(userId) // Explicitly excluded
+        ) {
+          queue.add(() => notify.user(activity, { ...options, skipUnsubscribedCheck: true, userId, user }));
+        }
+      }
+
+      await queue.onIdle();
```

Comment:
Two optimizations here:
- We accept a User ID instead of a full User. Since it's anyway reloaded later in the process, there's no benefit in providing a full model at this point.
- User a PQueue so that we only send 50 notifications in parallel. Should prevent memory exhaustion when loading 6000+ users and their collectives at once.